### DISCUSSION TITLE
Unable to open HID device in android (SELinux prevents it)

### DISCUSSION THREAD

**Discussion by tpkarras** [libusb/hidapi#461] [Q&A] [answered]
Hi. I'm new to hidapi so please forgive me.

I'm trying to open an input device in hidapi so I can change a setting that goes 1 or 0. I got it to attempt to open a device but pesky SELinux is preventing the device from being opened.

`2022-10-09 13:14:38.030 29678-29678/com.openrearultra I/m.openrearultra: type=1400 audit(0.0:2672): avc: denied { read } for name="usb" dev="tmpfs" ino=17549 scontext=u:r:untrusted_app:s0:c146,c257,c512,c768 tcontext=u:object_r:usb_device:s0 tclass=dir permissive=1 app=com.openrearultra
2022-10-09 13:14:38.030 29678-29678/com.openrearultra I/m.openrearultra: type=1400 audit(0.0:2673): avc: denied { open } for path="/dev/bus/usb" dev="tmpfs" ino=17549 scontext=u:r:untrusted_app:s0:c146,c257,c512,c768 tcontext=u:object_r:usb_device:s0 tclass=dir permissive=1 app=com.openrearultra
2022-10-09 13:14:38.030 29678-29678/com.openrearultra I/m.openrearultra: type=1400 audit(0.0:2674): avc: denied { read } for name="devices" dev="sysfs" ino=56340 scontext=u:r:untrusted_app:s0:c146,c257,c512,c768 tcontext=u:object_r:sysfs:s0 tclass=dir permissive=1 app=com.openrearultra
2022-10-09 13:14:38.030 29678-29678/com.openrearultra I/m.openrearultra: type=1400 audit(0.0:2675): avc: denied { open } for path="/sys/bus/usb/devices" dev="sysfs" ino=56340 scontext=u:r:untrusted_app:s0:c146,c257,c512,c768 tcontext=u:object_r:sysfs:s0 tclass=dir permissive=1 app=com.openrearultra
2022-10-09 13:14:38.030 29678-29678/com.openrearultra I/m.openrearultra: type=1400 audit(0.0:2676): avc: denied { read } for name="busnum" dev="sysfs" ino=70076 scontext=u:r:untrusted_app:s0:c146,c257,c512,c768 tcontext=u:object_r:sysfs:s0 tclass=file permissive=1 app=com.openrearultra`

yes... I'm using SELinux in permissive mode and got ro.debuggable set to 1.

is there anyway through the HID or I2C interface that I can access and change this setting. right now this is frustrating.

**Comment by Youw** [comment]
> is there anyway through the HID or I2C interface...

Interesting why you've mentioned I2C. Your device is a USB device, and HIDAPI only supports LIBUSB on Android.

Accessing a USB devices on Android is a known generic "issue". LIBUSB cannot access those directly like on other *nix platforms.
Take a look in details here: https://github.com/libusb/libusb/wiki/Android

TL;DR
---
Try [android/UsbManager](https://developer.android.com/reference/android/hardware/usb/UsbManager) + [`hid_libusb_wrap_sys_device`](https://github.com/libusb/hidapi/blob/master/libusb/hidapi_libusb.h).

**Reply by tpkarras** [reply]
It's an internal input device though. I'm trying to change a touchscreen device wake setting that's not visible through UsbManager and only visible through InputManager.

the touchscreen has an I2C driver but I don't know where to look.

if InputManager has a way of getting the file descriptor it'd be easy but alas. I'm stuck.

any ideas?

**Reply by Youw** [reply]
> not visible through UsbManager

So your device is not visible over android/UsbManager, but visible over HIDAPI?

**Reply by tpkarras** [reply]
I'm using the vendor and product IDs provided by InputDevice

**Comment by tpkarras** [comment]
update: rebuilt the library the proper way. still no luck. SELinux enforcement is disabled but when trying to open it, there's no pointer referring to the USB object, it's just a null reference.

Any advice would be appreciated in proper operation of HIDAPI.

**Reply by Youw** [reply]
Share the sample of C and Java code related to your HID device handling, if you want more detailed advice.

**Reply by tpkarras** [reply]
Okay. Here's some Java code (since I'm not using any C code).

I'm using JNA to reflect from the hidapi shared object, this is what I have for the function in my interface according to what HIDAPI's guide gives me.

`Pointer hid_open (short VID, short PID, Character serial);`

hid_open works (cause I can see the SELinux messages about accessing USB devices) but I don't know if it's a permission issue or something else entirely that's causing it to return null.

this is the code I have that takes an InputDevice's Product ID and Vendor ID and uses it to open an hidapi instance to the device. (in this case, an internal Cypress touch controller)

`      short vid = (short)inputDevice.getVendorId();
      short pid = (short)inputDevice.getProductId();
      Pointer HIDDevice = HIDInterface.INSTANCE.hid_open(vid, pid, null);`

Thanks for taking the time to help out, any solution provided would be helpful.

**Reply by Youw** [reply, marked as answer]
That's what I was trying to say in my first answer: `hid_open` will not work on Android.

What you need to do:
1) Get Android permission to open your USB device of interest:
```java
    usbManager = (UsbManager) getSystemService(Context.USB_SERVICE);
    HashMap<String, UsbDevice> deviceList = usbManager.getDeviceList();
    for (UsbDevice usbDevice : deviceList.values()) {
        usbManager.requestPermission(usbDevice, mPermissionIntent);
    }
```
2) Open that device (if user has granted the permission):
```java
    UsbDeviceConnection usbDeviceConnection = usbManager.openDevice(camDevice);
    int fileDescriptor = usbDeviceConnection.getFileDescriptor();
```
3) Open the device using [hid_libusb_wrap_sys_device](https://github.com/libusb/hidapi/blob/master/libusb/hidapi_libusb.h):
```java
    Pointer HIDDevice = HIDInterface.INSTANCE.hid_libusb_wrap_sys_device(fileDescriptor, -1);
```

**Reply by Youw** [reply]
P.S.: (modified) sample is taken from here: https://github.com/libusb/libusb/blob/master/android/README
